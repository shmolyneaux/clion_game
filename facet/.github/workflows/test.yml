name: Tests

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]
  merge_group:

jobs:
  code-quality:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run code quality checks
        shell: bash
        run: |
          just code-quality-ci

  test:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run tests
        shell: bash
        run: |
          just test-ci

  nostd:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run nostd tests
        shell: bash
        run: |
          just nostd-ci

  doc-tests:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run doc tests
        shell: bash
        run: |
          just doc-tests-ci

  miri:
    runs-on: depot-ubuntu-24.04-64

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-miri-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Run miri
        shell: bash
        run: |
          export CI=true
          just miri

  msrv:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - uses: taiki-e/install-action@cargo-hack

      - name: ✨ Check MSRV
        shell: bash
        run: |
          just msrv

  minimal-versions:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Check minimal versions
        shell: bash
        run: |
          # Install nightly for minimal-versions feature
          rustup toolchain install nightly
          # Generate lockfile with minimal versions
          cargo +nightly generate-lockfile -Z minimal-versions
          # Check with stable using minimal versions
          cargo check --workspace --all-features --locked --keep-going

  docs:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Check documentation
        shell: bash
        env:
          RUSTDOCFLAGS: -D warnings
        run: |
          just docs

  lockfile:
    runs-on: depot-ubuntu-24.04-4

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Check lockfile is updated
        shell: bash
        run: |
          cargo update --workspace --locked

  clippy:
    runs-on: depot-ubuntu-24.04-16

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-amd64
    permissions:
      security-events: write # to upload sarif results
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: Install SARIF tools
        shell: bash
        run: |
          cargo install clippy-sarif --locked
          cargo install sarif-fmt --locked

      - name: ✨ Run clippy with SARIF output
        shell: bash
        run: |
          cargo clippy --workspace --all-features --all-targets --message-format=json | clippy-sarif | tee clippy-results.sarif | sarif-fmt
        continue-on-error: true

      - name: Upload SARIF results
        uses: github/codeql-action/upload-sarif@v3
        with:
          sarif_file: clippy-results.sarif
          wait-for-processing: true

      - name: Report status
        shell: bash
        run: |
          cargo clippy --workspace --all-features --all-targets --keep-going -- -D warnings --allow deprecated

  test-macos:
    runs-on: depot-macos-latest
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - uses: taiki-e/install-action@v2
        with:
          tool: cargo-nextest

      - name: ✨ Run nextest tests (macOS)
        shell: bash
        run: |
          cargo nextest run --all-features

  test-windows:
    runs-on: depot-windows-2022
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - uses: taiki-e/install-action@v2
        with:
          tool: cargo-nextest

      - name: ✨ Run nextest tests (Windows)
        shell: bash
        run: |
          cargo nextest run --all-features

  coverage:
    runs-on: depot-ubuntu-24.04-32

    container:
      image: ghcr.io/facet-rs/facet-ci:latest-miri-amd64
    steps:
      - uses: actions/checkout@v4

      - uses: Swatinem/rust-cache@v2

      - name: ✨ Gather coverage
        shell: bash
        run: |
          cargo llvm-cov --no-report nextest
          cargo llvm-cov --no-report --doc
          mkdir coverage
          cargo llvm-cov report --doctests --lcov --output-path coverage/lcov.info

      - name: ✨ Publish to Coveralls
        uses: coverallsapp/github-action@master
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
