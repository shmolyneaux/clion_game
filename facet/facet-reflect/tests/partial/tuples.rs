use facet_reflect::Partial;
use facet_testhelpers::test;

#[cfg(not(miri))]
macro_rules! assert_snapshot {
    ($($tt:tt)*) => {
        insta::assert_snapshot!($($tt)*);
    };
}
#[cfg(miri)]
macro_rules! assert_snapshot {
    ($($tt:tt)*) => {
        /* no-op under miri */
    };
}

#[test]
fn build_empty_tuple() {
    // Test building ()
    let mut partial = Partial::alloc::<()>()?;
    partial.build()?;
}

#[test]
fn build_single_empty_tuple() {
    // Test building (())
    let mut partial = Partial::alloc::<((),)>()?;

    // Field 0 is of type ()
    partial.begin_nth_field(0)?;
    // Now we're working with type (), which has no fields
    partial.end()?;

    let single_empty = *partial.build()?;
    assert_eq!(single_empty, ((),));
}

#[test]
fn build_double_empty_tuple() {
    // Test building ((()),)
    let mut partial = Partial::alloc::<(((),),)>()?;

    // Field 0 is of type (())
    partial.begin_nth_field(0)?;

    // Now we're in (()) - field 0 is of type ()
    partial.begin_nth_field(0)?;
    // Now we're working with type (), which has no fields
    partial.end()?;

    // End the (()) field
    partial.end()?;

    let double_empty = *partial.build()?;
    assert_eq!(double_empty, (((),),));
}

#[test]
fn build_mixed_tuple() {
    // Test building (String, i32)
    let mut partial = Partial::alloc::<(String, i32)>()?;

    partial.begin_nth_field(0)?;
    partial.set("Hello".to_string())?;
    partial.end()?;

    partial.begin_nth_field(1)?;
    partial.set(42i32)?;
    partial.end()?;

    let mixed = *partial.build()?;
    assert_eq!(mixed, ("Hello".to_string(), 42));
}

#[test]
fn build_nested_tuple() {
    // Test building ((String, i32), bool)
    let mut partial = Partial::alloc::<((String, i32), bool)>()?;

    // Field 0 is of type (String, i32)
    partial.begin_nth_field(0)?;

    partial.begin_nth_field(0)?;
    partial.set("World".to_string())?;
    partial.end()?;

    partial.begin_nth_field(1)?;
    partial.set(99i32)?;
    partial.end()?;

    partial.end()?;

    // Field 1 is of type bool
    partial.begin_nth_field(1)?;
    partial.set(true)?;
    partial.end()?;

    let nested = *partial.build()?;
    assert_eq!(nested, (("World".to_string(), 99), true));
}

#[test]
fn test_issue_691_tuple_too_few_fields() {
    // This test verifies that issue #691 is fixed: attempting to build a tuple
    // with too few fields should return an error, not cause unsoundness.
    // The original issue showed that with the old Wip API, building a tuple
    // with insufficient fields could lead to accessing uninitialized memory.

    // Test case 1: 2-element tuple with only 1 field initialized
    let mut partial = Partial::alloc::<(String, String)>()?;
    partial.begin_nth_field(0)?;
    partial.set("a".to_string())?;
    partial.end()?;
    // Should fail because we didn't initialize the second field
    assert_snapshot!(partial.build().unwrap_err());
}

#[test]
fn test_issue_691_3_tuple_missing_field() {
    // Test case 2: 3-element tuple with only 2 fields initialized
    let mut partial = Partial::alloc::<(String, i32, bool)>()?;
    partial.begin_nth_field(0)?;
    partial.set("hello".to_string())?;
    partial.end()?;
    partial.begin_nth_field(1)?;
    partial.set(42)?;
    partial.end()?;
    // Should fail because we didn't initialize the third field
    assert_snapshot!(partial.build().unwrap_err());
}

#[test]
fn test_issue_691_nested_tuple_incomplete() {
    // Test case 3: Nested tuple with inner tuple not fully initialized
    let mut partial = Partial::alloc::<((String, i32), bool)>()?;
    partial.begin_nth_field(0)?;
    partial.begin_nth_field(0)?;
    partial.set("nested".to_string())?;
    partial.end()?;
    // We didn't set the i32 field of the inner tuple
    // The error should occur when we try to end the inner tuple frame
    assert_snapshot!(partial.end().unwrap_err());
}

#[test]
fn test_issue_691_valid_tuples() {
    // Test case 4: Empty tuple should work (no fields to initialize)
    let mut partial = Partial::alloc::<()>()?;
    let result = partial.build();
    assert!(result.is_ok(), "Building empty tuple should succeed");

    // Test case 5: Single-element tuple with field initialized should work
    let mut partial = Partial::alloc::<(String,)>()?;
    partial.begin_nth_field(0)?;
    partial.set("single".to_string())?;
    partial.end()?;
    let result = partial.build();
    assert!(
        result.is_ok(),
        "Building single-element tuple with field initialized should succeed"
    );
}
