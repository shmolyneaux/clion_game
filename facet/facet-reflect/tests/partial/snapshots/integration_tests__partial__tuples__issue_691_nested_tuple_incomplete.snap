---
source: facet-reflect/tests/partial/tuples.rs
expression: partial.end().unwrap_err()
---
Field '(String, i32)::1' was not initialized
